src/Features/CSharp/Portable/ConvertNamespace/ConvertNamespaceCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,76 @@
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertNamespace
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = PredefinedCodeRefactoringProviderNames.ConvertNamespace), Shared]
+    internal class ConvertNamespaceCodeRefactoringProvider : CodeRefactoringProvider
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public ConvertNamespaceCodeRefactoringProvider()
+        {
+        }
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var (document, span, cancellationToken) = context;
+            if (!span.IsEmpty)
+                return;
+
+            var position = span.Start;
+            var root = (CompilationUnitSyntax)await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(position);
+            var namespaceDecl = token.GetAncestor<BaseNamespaceDeclarationSyntax>();
+            if (namespaceDecl == null)
+                return;
+
+            if (!IsValidPosition(namespaceDecl, position))
+                return;
+
+            var optionSet = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var title =
+                ConvertNamespaceHelper.CanOfferUseBlockScoped(optionSet, namespaceDecl, forAnalyzer: false) ? CSharpFeaturesResources.Convert_to_block_scoped_namespace :
+                ConvertNamespaceHelper.CanOfferUseFileScoped(optionSet, root, namespaceDecl, forAnalyzer: false) ? CSharpFeaturesResources.Convert_to_file_scoped_namespace : null;
+            if (title == null)
+                return;
+
+            context.RegisterRefactoring(new MyCodeAction(
+                title, c => ConvertNamespaceHelper.ConvertAsync(document, namespaceDecl, c)));
+        }
+
+        private static bool IsValidPosition(BaseNamespaceDeclarationSyntax baseDeclaration, int position)
+        {
+            if (position < baseDeclaration.SpanStart)
+                return false;
+
+            if (baseDeclaration is FileScopedNamespaceDeclarationSyntax fileScopedNamespace)
+                return position <= fileScopedNamespace.SemicolonToken.Span.End;
+
+            if (baseDeclaration is NamespaceDeclarationSyntax namespaceDeclaration)
+                return position <= namespaceDeclaration.Name.Span.End;
+
+            throw ExceptionUtilities.UnexpectedValue(baseDeclaration.Kind());
+        }
+
+        private class MyCodeAction : CodeAction.DocumentChangeAction
+        {
+            public MyCodeAction(string title, Func<CancellationToken, Task<Document>> createChangedDocument)
+                : base(title, createChangedDocument, title)
```

Comment:
Should this be a non-localized equivalence key same as analyzers?

Reply:
I'll change. However the idea of a refactoring with an equivalence key is a bit wonky in the first place afaict :-)

Reply:
> the idea of a refactoring with an equivalence key is a bit wonky in the first place afaict :-)

That's why I commented that more as a question 😄